qiskit/transpiler/passes/scheduling/time_unit_conversion.py

```diff
@@ -135,3 +175,74 @@ def _unified(unit_set: Set[str]) -> str:
             return "SI"
 
         return "mixed"
+
+
+_DURATION_KIND_NAME = {
+    Duration.dt: "dt",
+    Duration.ns: "ns",
+    Duration.us: "us",
+    Duration.ms: "ms",
+    Duration.s: "s",
+}
+
+
+class _EvalDurationImpl(expr.ExprVisitor[float]):
+    """Evaluates the expression to a single float result.
+
+    If ``dt`` is provided or all durations are already in ``dt``, the result is in ``dt``.
+    Otherwise, the result will be in seconds, and all durations MUST be in wall-time (SI).
+    """
+
+    __slots__ = ("dt", "has_dt", "has_si")
+
+    def __init__(self, dt=None):
+        self.dt = dt if dt is not None else 1
+        self.has_dt = False
+        self.has_si = False
+
+    def in_cycles(self):
+        """Returns ``True`` if units are 'dt' after visit."""
+        return self.has_dt or self.dt != 1
+
+    def visit_value(self, node, /) -> float:
+        if isinstance(node.value, float):
+            return node.value
+        if isinstance(node.value, Duration.dt):
+            if self.has_si and self.dt == 1:
+                raise TranspilerError(
+                    "Fail to unify time units in delays. SI units "
+                    "and dt unit must not be mixed when dt is not supplied."
+                )
+            self.has_dt = True
+            return node.value[0]
+        if isinstance(node.value, Duration):
+            if self.has_dt and self.dt == 1:
+                raise TranspilerError(
+                    "Fail to unify time units in delays. SI units "
+                    "and dt unit must not be mixed when dt is not supplied."
+                )
+            self.has_si = True
+            if isinstance(node.value, Duration.s):
+                return node.value[0] / self.dt
+            from_unit = _DURATION_KIND_NAME.get(type(node.value))
+            return apply_prefix(node.value[0], from_unit) / self.dt
+        raise TranspilerError(f"invalid duration expression: {node}")
+
+    def visit_binary(self, node, /) -> float:
```

Comment:
This makes me realise that we added the float type but didn't add unary negation as an op.  No need to do it in this PR, just noting that we should add it for Qiskit 2.1.

Reply:
Ah that's a good point. Created issue #13966